### PR TITLE
Improvements and new features

### PR DESCRIPTION
**Pull request by aknysh**
## what
* Improvements and new features

## why
* `atmos terraform apply` made 100% compatible with `terraform apply`:
  - does not run `terraform plan`
  - `atmos terraform apply` does not use `planfile` (by default)
  - `atmos terraform apply` will ask for the confirmation to apply
  - `atmos terraform apply -auto-approve` will apply without asking for the confirmation to apply
  - The above behavior can be overridden. If `apply_auto_approve` config (or ENV var `ATMOS_COMPONENTS_TERRAFORM_APPLY_AUTO_APPROVE `) is set to `true`, `atmos terraform apply` will not ask for the confirmation to apply regardless of whether the `-auto-approve` is specified or not

* `atmos terraform deploy`:
  - behaves like `atmos terraform apply -auto-approve`
  - does not run `terraform plan`
  - If `deploy_run_init` config (or ENV var `ATMOS_COMPONENTS_TERRAFORM_DEPLOY_RUN_INIT`, or `--deploy-run-init` command-line argument) is set to `true`, `atmos terraform deploy` will first run `terraform init` to initialize modules. Otherwise, it will not run `terraform init`, and `atmos terraform init` will need to be run first at least once

  ```
  ATMOS_COMPONENTS_TERRAFORM_DEPLOY_RUN_INIT=false atmos terraform deploy top-level-component1 -s tenant1-ue2-dev
  ATMOS_COMPONENTS_TERRAFORM_DEPLOY_RUN_INIT=true atmos terraform deploy top-level-component1 -s tenant1-ue2-dev
  ```

  ```
  atmos terraform deploy top-level-component1 -s tenant1/ue2/dev --deploy-run-init=false
  
  Using command line argument '--deploy-run-init=false'
  ```

  ```
  ATMOS_COMPONENTS_TERRAFORM_DEPLOY_RUN_INIT=false atmos terraform deploy top-level-component1 -s tenant1/ue2/dev --deploy-run-init=true
  
  Found ENV var ATMOS_COMPONENTS_TERRAFORM_DEPLOY_RUN_INIT=false
  Using command line argument '--deploy-run-init=true'
  ```

* Log level control:
  - If ENV var `ATMOS_LOGS_VERBOSE ` is set to `false` (default), the CLI will not print all the information about the config, stacks and component (and how it searches them), but will just output the essential info for each command

    ```
    ATMOS_LOGS_VERBOSE=true atmos terraform deploy top-level-component1 -s tenant1-ue2-dev
    ```

* Check if terraform commands require user interaction, but running in a scripted environment (where a `tty` is not attached or `stdin` is not attached)

* Use logical `context` to generate terraform workspaces and file names for `planfile` and `varfile`:
  - Instead of using the physical paths derived from the provided stack names to generate terraform workspaces and the names of  `planfile` and `varfile` (which the old variant-based `atmos` did as well), find the logical `context` for the specified component in the specified stack (from the deep-merged vars). The logical `context` consist of `tenant` (if used), `environment` and `stage`, and is controlled by the `name_pattern: "{tenant}-{environment}-{stage}"` CLI config
  - This makes it completely independent of the folder structure where the YAML stack config files are defined - don't use the physical structure, look at the logical structure
  - Bonus:
    - You will be able move around the top-level YAML config files (that are not imported into other stacks) to any folder at any level (under `stacks` parent folder) "on-the-fly" , or change the names of the folders, WITHOUT needing to change anything - terraform workspace remains the same, the CLI will find the context from the deep-merged vars, and it just works
    - You can also move around the YAML config files even if they are already imported into other stacks, you will just need to update the `import` paths, but all CLI commands will continue working without making any changes to terraform state

### CODE REVIEW

**Review by Nuru — CHANGES_REQUESTED** [review]
- `-auto-approve=false` and/or `-no-auto-approve` should be available to override a config of `apply_auto_approve: true`
- Default should be for `atmos terraform deploy` to run `init` first, and there likewise should be command line options to override whatever is set by `deploy_run_init` and/or ENV vars.
- `atmos` should exit with an error if it is configured to ask for approval but `stdin` is not a terminal.

File: internal/config/config.go
Comment (aknysh) [reply internal/config/config.go]: it's a wording issue, fixed it.
The CLI first checks folders if it can find the files provided by the stack name using Globs.
If it does not find anything, it will search for config files using `context` and the provided `stack_name_pattern`.
In this case, the file is in the top-level folder, but the Glob does not know it since it finds it anyway.

Updated the messages to be simply:

```
The stack '%s' matches the stack config file %s
or
The stack '%s' matches the stack name pattern '%s'
```

**Comment by aknysh** [comment]
> * `-auto-approve=false` and/or `-no-auto-approve` should be available to override a config of `apply_auto_approve: true`
> * Default should be for `atmos terraform deploy` to run `init` first, and there likewise should be command line options to override whatever is set by `deploy_run_init` and/or ENV vars.
> * `atmos` should exit with an error if it is configured to ask for approval but `stdin` is not a terminal.

1. We already have a command-line argument `-auto-approve` which can be used with all terraform commands that support it. The `apply_auto_approve` config can be used to override it (to always set it to `true` regardless of any other command-line arguments) - it's `false` by default and has no effect. We don't need to create another command-line argument to override the override (too complex and difficult to explain why).
2. Added command-line argument `--deploy-run-init`
3. Added check for `stdin` not attached

**Review by osterman — APPROVED** [review]

**Comment by Nuru** [comment]
> > * `-auto-approve=false` and/or `-no-auto-approve` should be available to override a config of `apply_auto_approve: true`
> 
>     1. We already have a command-line argument `-auto-approve` which can be used with all terraform commands that support it. The `apply_auto_approve` config can be used to override it (to always set it to `true` regardless of any other command-line arguments) - it's `false` by default and has no effect. We don't need to create another command-line argument to override the override (too complex and difficult to explain why).
> 

Cobra/pflags would support `--auto-approve=false` if `--auto-approve` were a boolean flag. May Unix utilities support the `--[no-]boolean-flag`. It's not that hard to explain. The command line options take precedence over other configuration options like config files and ENV vars. You want to have a command line option that specified the default behavior so that you can force it even when other configuration tries to change it. It is pretty standard practice.

**Review by Nuru — APPROVED** [review]
I disagree with @aknysh  about `-auto-approve=false` (see [here](https://github.com/cloudposse/atmos/pull/58#issuecomment-935379958)) but will not block the PR because if it.

**Pull request merged**
